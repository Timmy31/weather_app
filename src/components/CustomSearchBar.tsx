import React from 'react';
import {
  View,
  TextInput,
  TouchableOpacity,
  StyleSheet,
  FlatList,
  Text,
} from 'react-native';
import Ionicons from 'react-native-vector-icons/Ionicons';
import { CityProps } from '../types';

interface CustomSearchBarProps {
  fetchSuggestionList: (text: string) => void;
  handleSearch: (text: string) => void;
  onPressSearch: () => void;
  suggestions: CityProps[];
}

export const CustomSearchBar: React.FC<CustomSearchBarProps> = ({
  fetchSuggestionList,
  handleSearch,
  onPressSearch,
  suggestions,
}) => {
  const renderItem = ({item}: {item: CityProps}) => (
    <TouchableOpacity
      style={styles.suggestionItem}
      onPress={() => {
        handleSearch(item.name);
        onPressSearch();
      }}>
      <Text>
        {item.name}, {item.country}
      </Text>
    </TouchableOpacity>
  );

  return (
    <View style={styles.container}>
      <TextInput
        style={styles.input}
        placeholder="Search location..."
        onChangeText={searchCity => {
          fetchSuggestionList(searchCity);
          handleSearch(searchCity);
        }}
        onSubmitEditing={onPressSearch}
      />
      <TouchableOpacity style={styles.iconContainer} onPress={onPressSearch}>
        <Ionicons name="search" size={20} color="#999" />
      </TouchableOpacity>
      <FlatList
        style={styles.suggestionList}
        data={suggestions}
        renderItem={renderItem}
        keyExtractor={item => item.id.toString()}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    position: 'relative',
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: '#fff',
    borderRadius: 5,
    padding: 8,
    zIndex: 9,
  },
  input: {
    flex: 1,
    marginHorizontal: 16,
  },
  iconContainer: {
    padding: 8,
  },
  suggestionList: {
    position: 'absolute',
    top: 60,
    left: 0,
    right: 0,
    backgroundColor: '#fff',
    elevation: 3,
    maxHeight: 250,
    overflow: 'scroll',
  },
  suggestionItem: {
    padding: 10,
    borderBottomWidth: 1,
    borderBottomColor: '#eee',
  },
});
