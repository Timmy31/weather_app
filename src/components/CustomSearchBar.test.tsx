import React from 'react';
import { create } from 'react-test-renderer';
import { act } from 'react-test-renderer';
import { CustomSearchBar } from './CustomSearchBar'; // Adjust the path based on your project structure

// Mocking the react-native-vector-icons library
jest.mock('react-native-vector-icons/Ionicons', () => 'Ionicons');

describe('CustomSearchBar Component', () => {
  const mockFetchSuggestionList = jest.fn();
  const mockHandleSearch = jest.fn();
  const mockOnPressSearch = jest.fn();

  const suggestions = [
    { id: 1, name: 'City1', country: 'Country1', region: 'Region1', lat: '123234', log: '12334' },
    { id: 2, name: 'City2', country: 'Country2', region: 'Region2', lat: '123232', log: '12332' },
  ];

  it('renders correctly', () => {
    const tree = create(
      <CustomSearchBar
        fetchSuggestionList={mockFetchSuggestionList}
        handleSearch={mockHandleSearch}
        onPressSearch={mockOnPressSearch}
        suggestions={suggestions}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls fetchSuggestionList and handleSearch on text input change', () => {
    const component = create(
      <CustomSearchBar
        fetchSuggestionList={mockFetchSuggestionList}
        handleSearch={mockHandleSearch}
        onPressSearch={mockOnPressSearch}
        suggestions={suggestions}
      />
    );

    const input = component.root.findByProps({ testID: 'search-input' });

    act(() => {
      input.props.onChangeText('City');
    });

    expect(mockFetchSuggestionList).toHaveBeenCalledWith('City');
    expect(mockHandleSearch).toHaveBeenCalledWith('City');
  });

  it('calls onPressSearch when the search button is pressed', () => {
    const component = create(
      <CustomSearchBar
        fetchSuggestionList={mockFetchSuggestionList}
        handleSearch={mockHandleSearch}
        onPressSearch={mockOnPressSearch}
        suggestions={suggestions}
      />
    );

    const button = component.root.findByProps({ testID: 'search-button' });

    act(() => {
      button.props.onPress();
    });

    expect(mockOnPressSearch).toHaveBeenCalled();
  });

  it('calls handleSearch and onPressSearch when a suggestion is pressed', () => {
    const component = create(
      <CustomSearchBar
        fetchSuggestionList={mockFetchSuggestionList}
        handleSearch={mockHandleSearch}
        onPressSearch={mockOnPressSearch}
        suggestions={suggestions}
      />
    );

    const suggestion = component.root.findByProps({ testID: 'search-suggestion-list-1' });

    act(() => {
      suggestion.props.onPress();
    });

    expect(mockHandleSearch).toHaveBeenCalledWith('City1');
    expect(mockOnPressSearch).toHaveBeenCalled();
  });
});
