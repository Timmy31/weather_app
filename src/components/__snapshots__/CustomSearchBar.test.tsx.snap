// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomSearchBar Component renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "borderRadius": 5,
      "flexDirection": "row",
      "padding": 8,
      "position": "relative",
      "zIndex": 9,
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    onSubmitEditing={[MockFunction]}
    placeholder="Search location..."
    style={
      {
        "flex": 1,
        "marginHorizontal": 16,
      }
    }
    testID="search-input"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
        "padding": 8,
      }
    }
    testID="search-button"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#999",
            "fontSize": 20,
          },
          undefined,
          {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
      testID="search-icon"
    >
      î¹ 
    </Text>
  </View>
  <RCTScrollView
    data={
      [
        {
          "country": "Country1",
          "id": 1,
          "name": "City1",
        },
        {
          "country": "Country2",
          "id": 2,
          "name": "City2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    style={
      {
        "backgroundColor": "#ffffff",
        "elevation": 3,
        "left": 0,
        "maxHeight": 250,
        "overflow": "scroll",
        "position": "absolute",
        "right": 0,
        "top": 60,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "borderBottomColor": "#e1e8ee",
              "borderBottomWidth": 1,
              "opacity": 1,
              "padding": 10,
            }
          }
          testID="search-suggestion-list"
        >
          <Text>
            City1
            , 
            Country1
          </Text>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "borderBottomColor": "#e1e8ee",
              "borderBottomWidth": 1,
              "opacity": 1,
              "padding": 10,
            }
          }
          testID="search-suggestion-list"
        >
          <Text>
            City2
            , 
            Country2
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
